stl/inc/xutility

```diff
@@ -2668,9 +2744,223 @@ namespace ranges {
+                        _Where = static_cast<_Se&&>(_Last);
+                    } else {
+                        (*this)(_Where, _Delta);
+                    }
+                    return _Off - _Delta;
+                }
+
+                (*this)(_Where, _Off);
+                return 0;
+            } else {
+                // performance note: develop unwrapping technology for (i, n, s)?
+                if constexpr (bidirectional_iterator<_It>) {
+                    for (; _Off < 0 && _Where != _Last; ++_Off) {
+                        --_Where;
+                    }
+                } else {
+                    _STL_ASSERT(_Off >= 0, "negative advance of non-bidirectional iterator");
+                }
+
+                for (; _Off > 0 && _Where != _Last; --_Off) {
+                    ++_Where;
+                }
+
+                return _Off;
+            }
+        }
+    };
+
+    // VARIABLE ranges::advance
+    inline constexpr _Advance_fn advance{_Not_quite_object::_Construct_tag{}};
+
+    // CLASS ranges::_Distance_fn
+    class _Distance_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        template <input_or_output_iterator _It, sentinel_for<_It> _Se>
+        _NODISCARD constexpr iter_difference_t<_It> operator()(_It _First, _Se _Last) const {
+            if constexpr (sized_sentinel_for<_Se, _It>) {
+                return _Last - _First;
+            } else {
+                _Adl_verify_range(_First, _Last);
+                auto _UFirst                  = _Get_unwrapped(static_cast<_It&&>(_First));
+                const auto _ULast             = _Get_unwrapped(static_cast<_Se&&>(_Last));
+                iter_difference_t<_It> _Count = 0;
+                for (; _UFirst != _ULast; ++_UFirst) {
+                    ++_Count;
+                }
+
+                return _Count;
+            }
+        }
+
+        template <range _Rng>
+        _NODISCARD constexpr range_difference_t<_Rng> operator()(_Rng&& _Val) const {
+            if constexpr (sized_range<_Rng>) {
+                return static_cast<range_difference_t<_Rng>>(_RANGES size(_Val));
+            } else {
+                return (*this)(_RANGES begin(_Val), _RANGES end(_Val));
+            }
+        }
+    };
+
+    // VARIABLE ranges::distance
+    inline constexpr _Distance_fn distance{_Not_quite_object::_Construct_tag{}};
+
+    // CLASS ranges::_Next_fn
+    class _Next_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        template <input_or_output_iterator _It>
+        _NODISCARD constexpr _It operator()(_It _Where) const {
+            ++_Where;
+            return _Where;
+        }
+
+        template <input_or_output_iterator _It>
+        _NODISCARD constexpr _It operator()(_It _Where, const iter_difference_t<_It> _Off) const {
+            _RANGES advance(_Where, _Off);

```

Comment:
Out of curiosity, could you elaborate why there is no assert here despite the one in `_Prev_fn`? Shouldn't each be equivalent to calling the other one with `-_Off`?

Reply:
`next(i, n)` and `prev(i, n)` are equivalent to `advance(i, n)` and `advance(i, -n)` , respectively (ignoring for the moment that `advance` takes a reference and `next`/`prev` accept an and return a value). The crucial difference between `next` and `prev` is that `prev` evaluates `-n`, which potentially has undefined behavior when `n` is the most-negative value of a signed integral type. The assert in `prev` is there to check for that particular value. `next`, conversely, always has defined behavior.